samples/generic-synthetic-broken-links/brokenLinkChecker.js

```diff
@@ -0,0 +1,136 @@
+    const anchors = Array.from(document.querySelectorAll('a'));
+    return anchors.map(anchor => anchor.href);
+  });
+
+  // Filter out duplicate links and remove trailing "#"
+  const filteredLinks = [...new Set(
+    links.map(link => link.toString().replace(/#$/, ''))
+  )].filter(link => link.startsWith('http') || link.startsWith('https'));
+
+  return filteredLinks;
+}
+
+
+/**
+ * Checks the status of `maxNumberOfFollowedLinks` links scraped on `startUrl`
+ * webpage leveraging Puppeteer.
+ *
+ * @param {string} startUrl - The starting URL to check the links on.
+ * @param {number} maxNumberOfFollowedLinks - The number of links to check.
+ * @param {number} maxTimeout - The maximum timeout for each link request in ms.
+ * @returns {Promise<Array<Object>>} - Array of link status results.
+ */
+async function checkLinks(startUrl, maxNumberOfFollowedLinks, maxTimeout) {
+  const browser = await puppeteer.launch({
+    headless: "new"
+  })
+  let page = await browser.newPage();
+  const linksToReturn = [];
+  const linksToCheck = [];
+
+  // load initial page, return any error in LinkResult format
+  try {
+    // Navigate to the URL and wait for the page to finish loading
+    const parentResponse = await page.goto(
+      startUrl,
+      { waitUntil: 'load', timeout: maxTimeout });
+    linksToReturn.push({
+      targetUrl: startUrl,
+      statusCode: parentResponse.status(),
+    });
+
+  } catch (error) {
+    const error_message =
+      'Could not successfully navigate to startUrl. Error: ' + error.message;
+    throw new Error(error_message);
+  };
+
+  // retrive all links from page
+  try {
+    linksToCheck.push(...await retrieveLinks(page));
+  } catch (error) {
+    const error_message =
+      'Could not successfully retrieve links. Error: ' + error.message;
+    throw new Error(error_message);
+  }
+
+  linksToCheck.splice(maxNumberOfFollowedLinks);
+
+  page = await browser.newPage();
+  page.setCacheEnabled(false); // prevents 304 errors
+  page.setDefaultNavigationTimeout(maxTimeout);
+
+  for (const targetLink of linksToCheck) {
+    try {
+      const response = await page.goto(
+        targetLink,
+        { waitUtil: 'load', timeout: maxTimeout }
+      );
+
+      linksToReturn.push({
+        targetUrl: targetLink,
+        statusCode: response.status(),
+      });
+    } catch (error) {
+      linksToReturn.push({
+        targetUrl: targetLink,
+        statusCode: null
+      });
+    }
+  }
+
+  await browser.close();
+  return linksToReturn;
+}
+
+
+/*
+ * crawls links and returns a dictionary mapping failing (non2xx) links' target
+ * URLs to their respective status codes.
+ *
+ * @param {string} startUrl - The starting URL to check the links on.
+ * @param {number} maxNumberOfFollowedLinks - The number of links to check.
+ * @param {number} maxTimeout - The maximum timeout for each link request in ms
+ */
+async function runBrokenLinks(startUrl, maxNumberOfFollowedLinks = 30, maxTimeout = 5000) {
```

Comment:
Thinking about the output of this function, this is supposed to be called by runSyntheticHandler. This means that that it should either exit gracefully, or it should throw an error. In light of that, I think that we should create this dictionary, and then throw an error if there's anything that's non-200 that includes which links failed. Think about alternatives as well for what that message could be.

Reply:
Ok thinking about this a bit more after ive read the whole thing.

I think that this would make sense for the reader of this function to 
1) Get linksResults
2) Filter results to get all links that are non-200
3) Throw an error if necessary.

To that, I would say that you should construct that error in  createNon2xxLinksDictionary in this function instead, otherwise the source of the error will be harder to track.